File: lj_core.py
from __future__ import annotations

import math
from dataclasses import dataclass
from itertools import product

import numpy as np


@dataclass
class NeighborState:
    use_cutoff: bool
    cutoff_radius: float | None
    use_cell_list: bool
    cell_size: float | None
    n_cells: int | None = None
    cells: dict[tuple[int, ...], list[int]] | None = None
    particle_cells: list[tuple[int, ...]] | None = None


def validate_common_parameters(
    *,
    temperature: float,
    n_particles: int,
    n_steps: int,
    delta_max: float,
    box_size: float,
    epsilon: float,
    sigma: float,
) -> None:
    """単箱 MC / GEMC で共通の主要パラメータを検証する。"""
    if temperature <= 0:
        raise ValueError(f"T must be positive, got {temperature}")
    if n_particles <= 0:
        raise ValueError(f"N must be positive, got {n_particles}")
    if n_steps <= 0:
        raise ValueError(f"N_STEPS must be positive, got {n_steps}")
    if delta_max <= 0:
        raise ValueError(f"DELTA_MAX must be positive, got {delta_max}")
    if box_size <= 0:
        raise ValueError(f"BOX_SIZE must be positive, got {box_size}")
    if epsilon < 0:
        raise ValueError(f"EPSILON must be non-negative, got {epsilon}")
    if sigma <= 0:
        raise ValueError(f"SIGMA must be positive, got {sigma}")


def exclusion_volume_per_particle(min_distance: float, dim: int) -> float:
    """初期配置チェックに使う 1 粒子あたりの排他体積を返す。"""
    radius = min_distance / 2.0
    if dim == 2:
        return math.pi * radius * radius
    if dim == 3:
        return (4.0 / 3.0) * math.pi * radius**3
    raise ValueError(f"Unsupported dimension: {dim}")


def validate_initial_density(
    *,
    n_particles: int,
    box_size: float,
    min_distance: float,
    dim: int,
    packing_limit: float = 0.8,
) -> None:
    """ランダム初期配置が極端に高密度でないかを確認する。"""
    required_measure = n_particles * exclusion_volume_per_particle(min_distance, dim)
    available_measure = box_size**dim
    if required_measure > packing_limit * available_measure:
        measure_name = "area" if dim == 2 else "volume"
        raise ValueError(
            f"Density too high: N={n_particles} particles with min_distance={min_distance:.3f} "
            f"require {measure_name} ~{required_measure:.3f}, but box has {available_measure:.3f}. "
            "Increase BOX_SIZE or decrease N."
        )


def normalize_neighbor_options(
    *,
    box_size: float,
    use_cutoff: bool,
    cutoff_radius: float | None,
    use_cell_list: bool,
    cell_size: float | None,
) -> tuple[bool, float | None, bool, float | None]:
    """cutoff と cell list の有効設定を正規化する。"""
    effective_use_cutoff = bool(use_cutoff)
    effective_cutoff = cutoff_radius
    effective_use_cell_list = bool(use_cell_list and use_cutoff)
    effective_cell_size = cell_size

    if effective_use_cutoff:
        if effective_cutoff is None:
            raise ValueError("cutoff_radius must be provided when use_cutoff=True.")
        if effective_cutoff <= 0:
            raise ValueError(f"cutoff_radius must be positive, got {effective_cutoff}")
        if effective_cutoff >= 0.5 * box_size:
            raise ValueError(
                f"cutoff_radius must be smaller than half the box size. "
                f"Got cutoff_radius={effective_cutoff}, box_size={box_size}."
            )
    else:
        effective_cutoff = None

    if effective_use_cell_list:
        if effective_cell_size is None:
            effective_cell_size = effective_cutoff
        if effective_cell_size is None or effective_cell_size <= 0:
            raise ValueError(f"cell_size must be positive, got {effective_cell_size}")
    else:
        effective_cell_size = None

    return effective_use_cutoff, effective_cutoff, effective_use_cell_list, effective_cell_size


def minimum_image(delta: np.ndarray, box_size: float) -> np.ndarray:
    """Minimum Image Convention を適用した差ベクトルを返す。"""
    return delta - box_size * np.round(delta / box_size)


def wrap_positions(pos: np.ndarray, box_size: float) -> np.ndarray:
    """周期境界条件で座標を箱内に折り返す。"""
    return pos % box_size


def init_positions(
    box_size: float,
    n_particles: int,
    min_distance: float,
    rng: np.random.Generator,
    *,
    dim: int,
    max_trials: int = 10_000,
    max_global_retries: int = 1_000,
) -> np.ndarray:
    """粒子をランダムに初期配置し、詰まった場合は全体を作り直す。"""
    min_distance_sq = min_distance**2

    for _ in range(max_global_retries):
        pos = np.empty((n_particles, dim), dtype=float)

        for idx in range(n_particles):
            for _ in range(max_trials):
                candidate = rng.uniform(0.0, box_size, size=dim)
                if idx == 0:
                    pos[idx] = candidate
                    break

                delta = candidate - pos[:idx]
                delta = minimum_image(delta, box_size)
                distances_sq = np.sum(delta * delta, axis=1)
                if np.all(distances_sq >= min_distance_sq):
                    pos[idx] = candidate
                    break
            else:
                break
        else:
            return pos

    raise ValueError(
        f"Failed to place {n_particles} particles with min_distance={min_distance:.3f} "
        f"in box_size={box_size} after {max_global_retries} attempts. "
        "Density may be too high."
    )


def _cell_index(point: np.ndarray, box_size: float, n_cells: int) -> tuple[int, ...]:
    scaled = np.floor((point / box_size) * n_cells).astype(int) % n_cells
    return tuple(int(x) for x in scaled)


def build_neighbor_state(
    pos: np.ndarray,
    box_size: float,
    *,
    use_cutoff: bool = False,
    cutoff_radius: float | None = None,
    use_cell_list: bool = False,
    cell_size: float | None = None,
) -> NeighborState:
    """cutoff / cell list 設定に応じた近傍探索状態を構築する。"""
    effective_use_cutoff, effective_cutoff, effective_use_cell_list, effective_cell_size = normalize_neighbor_options(
        box_size=box_size,
        use_cutoff=use_cutoff,
        cutoff_radius=cutoff_radius,
        use_cell_list=use_cell_list,
        cell_size=cell_size,
    )

    state = NeighborState(
        use_cutoff=effective_use_cutoff,
        cutoff_radius=effective_cutoff,
        use_cell_list=effective_use_cell_list,
        cell_size=effective_cell_size,
    )

    if not effective_use_cell_list:
        return state

    assert effective_cell_size is not None
    n_cells = max(1, int(box_size / effective_cell_size))
    state.n_cells = n_cells
    state.cells = {}
    state.particle_cells = []

    for idx, point in enumerate(pos):
        cell = _cell_index(point, box_size, n_cells)
        state.particle_cells.append(cell)
        state.cells.setdefault(cell, []).append(idx)

    return state


def rebuild_neighbor_state(state: NeighborState | None, pos: np.ndarray, box_size: float) -> NeighborState | None:
    """同じ設定のまま近傍探索状態を再構築する。"""
    if state is None:
        return None
    return build_neighbor_state(
        pos,
        box_size,
        use_cutoff=state.use_cutoff,
        cutoff_radius=state.cutoff_radius,
        use_cell_list=state.use_cell_list,
        cell_size=state.cell_size,
    )


def update_neighbor_state_after_move(
    state: NeighborState | None,
    idx: int,
    old_pos: np.ndarray,
    new_pos: np.ndarray,
    box_size: float,
) -> None:
    """1 粒子変位後に cell list のみを差分更新する。"""
    if state is None or not state.use_cell_list or state.cells is None or state.particle_cells is None or state.n_cells is None:
        return

    old_cell = _cell_index(old_pos, box_size, state.n_cells)
    new_cell = _cell_index(new_pos, box_size, state.n_cells)
    if old_cell == new_cell:
        return

    state.cells[old_cell].remove(idx)
    if not state.cells[old_cell]:
        del state.cells[old_cell]
    state.cells.setdefault(new_cell, []).append(idx)
    state.particle_cells[idx] = new_cell


def _neighbor_cells(cell: tuple[int, ...], n_cells: int) -> list[tuple[int, ...]]:
    neighbors: set[tuple[int, ...]] = set()
    for offset in product([-1, 0, 1], repeat=len(cell)):
        neighbors.add(tuple((cell[d] + offset[d]) % n_cells for d in range(len(cell))))
    return list(neighbors)


def _candidate_indices_for_particle(
    pos: np.ndarray,
    i: int,
    box_size: float,
    neighbor_state: NeighborState | None,
) -> list[int]:
    if neighbor_state is None or not neighbor_state.use_cell_list or neighbor_state.cells is None or neighbor_state.particle_cells is None or neighbor_state.n_cells is None:
        return list(range(len(pos)))

    cell = neighbor_state.particle_cells[i]
    candidates: list[int] = []
    for neighbor in _neighbor_cells(cell, neighbor_state.n_cells):
        candidates.extend(neighbor_state.cells.get(neighbor, []))
    return candidates


def _candidate_indices_for_point(
    candidate: np.ndarray,
    pos: np.ndarray,
    box_size: float,
    neighbor_state: NeighborState | None,
) -> list[int]:
    if neighbor_state is None or not neighbor_state.use_cell_list or neighbor_state.cells is None or neighbor_state.n_cells is None:
        return list(range(len(pos)))

    cell = _cell_index(candidate, box_size, neighbor_state.n_cells)
    candidates: list[int] = []
    for neighbor in _neighbor_cells(cell, neighbor_state.n_cells):
        candidates.extend(neighbor_state.cells.get(neighbor, []))
    return candidates


def lj_pair_energy(
    r: float,
    epsilon: float,
    sigma: float,
    *,
    use_cutoff: bool = False,
    cutoff_radius: float | None = None,
) -> float:
    """Lennard-Jones 12-6 ポテンシャル。"""
    if r < 1e-10:
        return 0.0
    if use_cutoff and cutoff_radius is not None and r >= cutoff_radius:
        return 0.0
    sr = sigma / r
    sr6 = sr**6
    sr12 = sr6**2
    return 4.0 * epsilon * (sr12 - sr6)


def calc_energy(
    pos: np.ndarray,
    i: int,
    box_size: float,
    epsilon: float,
    sigma: float,
    *,
    neighbor_state: NeighborState | None = None,
) -> float:
    """粒子 i と他粒子の LJ 相互作用エネルギー総和を返す。"""
    energy = 0.0

    for j in _candidate_indices_for_particle(pos, i, box_size, neighbor_state):
        if i == j:
            continue
        delta = minimum_image(pos[i] - pos[j], box_size)
        r = float(np.linalg.norm(delta))
        energy += lj_pair_energy(
            r,
            epsilon,
            sigma,
            use_cutoff=bool(neighbor_state and neighbor_state.use_cutoff),
            cutoff_radius=neighbor_state.cutoff_radius if neighbor_state else None,
        )

    return energy


def total_energy(
    pos: np.ndarray,
    box_size: float,
    epsilon: float,
    sigma: float,
    *,
    neighbor_state: NeighborState | None = None,
    use_cutoff: bool = False,
    cutoff_radius: float | None = None,
    use_cell_list: bool = False,
    cell_size: float | None = None,
) -> float:
    """全粒子系の全エネルギーを返す。"""
    state = neighbor_state
    if state is None and (use_cutoff or use_cell_list):
        state = build_neighbor_state(
            pos,
            box_size,
            use_cutoff=use_cutoff,
            cutoff_radius=cutoff_radius,
            use_cell_list=use_cell_list,
            cell_size=cell_size,
        )
    return sum(calc_energy(pos, i, box_size, epsilon, sigma, neighbor_state=state) for i in range(len(pos))) / 2.0


def insertion_energy(
    pos: np.ndarray,
    candidate: np.ndarray,
    box_size: float,
    epsilon: float,
    sigma: float,
    *,
    neighbor_state: NeighborState | None = None,
) -> float:
    """候補位置に 1 粒子を挿入したときの相互作用エネルギーを返す。"""
    if len(pos) == 0:
        return 0.0

    energy = 0.0
    for idx in _candidate_indices_for_point(candidate, pos, box_size, neighbor_state):
        delta = minimum_image(candidate - pos[idx], box_size)
        r = float(np.linalg.norm(delta))
        energy += lj_pair_energy(
            r,
            epsilon,
            sigma,
            use_cutoff=bool(neighbor_state and neighbor_state.use_cutoff),
            cutoff_radius=neighbor_state.cutoff_radius if neighbor_state else None,
        )
    return energy

File: lennard_jones_mc_rs/src/lj_core.rs
//! Lennard-Jones 12-6 ポテンシャルと周期境界条件の共通計算
//! Python lj_core.py の Rust 移植

use rand::Rng;
use std::collections::{HashMap, HashSet};

/// 3次元ベクトル
pub type Vec3 = [f64; 3];

/// 近傍探索状態（cutoff / cell list）
#[derive(Clone)]
pub struct NeighborState {
    pub use_cutoff: bool,
    pub cutoff_radius: Option<f64>,
    pub use_cell_list: bool,
    pub cell_size: Option<f64>,
    pub n_cells: Option<i32>,
    pub cells: Option<HashMap<(i32, i32, i32), Vec<usize>>>,
    pub particle_cells: Option<Vec<(i32, i32, i32)>>,
}

impl NeighborState {
    pub fn new(
        use_cutoff: bool,
        cutoff_radius: Option<f64>,
        use_cell_list: bool,
        cell_size: Option<f64>,
    ) -> Self {
        Self {
            use_cutoff,
            cutoff_radius,
            use_cell_list,
            cell_size,
            n_cells: None,
            cells: None,
            particle_cells: None,
        }
    }
}

/// Minimum Image Convention
#[inline]
pub fn minimum_image(delta: Vec3, box_size: f64) -> Vec3 {
    [
        delta[0] - box_size * (delta[0] / box_size).round(),
        delta[1] - box_size * (delta[1] / box_size).round(),
        delta[2] - box_size * (delta[2] / box_size).round(),
    ]
}

/// 周期境界で座標を箱内に折り返す
#[inline]
pub fn wrap_positions(pos: Vec3, box_size: f64) -> Vec3 {
    [
        pos[0].rem_euclid(box_size),
        pos[1].rem_euclid(box_size),
        pos[2].rem_euclid(box_size),
    ]
}

/// 2点間の差ベクトル（minimum image）
#[inline]
pub fn delta_minimum_image(a: Vec3, b: Vec3, box_size: f64) -> Vec3 {
    let d = [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
    minimum_image(d, box_size)
}

/// ベクトルのノルム
#[inline]
pub fn norm(v: Vec3) -> f64 {
    (v[0] * v[0] + v[1] * v[1] + v[2] * v[2]).sqrt()
}

/// Lennard-Jones 12-6 ペアエネルギー
#[inline]
pub fn lj_pair_energy(
    r: f64,
    epsilon: f64,
    sigma: f64,
    use_cutoff: bool,
    cutoff_radius: Option<f64>,
) -> f64 {
    if r < 1e-10 {
        return 0.0;
    }
    if use_cutoff {
        if let Some(rc) = cutoff_radius {
            if r >= rc {
                return 0.0;
            }
        }
    }
    let sr = sigma / r;
    let sr6 = sr * sr * sr * sr * sr * sr;
    let sr12 = sr6 * sr6;
    4.0 * epsilon * (sr12 - sr6)
}

fn cell_index(point: Vec3, box_size: f64, n_cells: i32) -> (i32, i32, i32) {
    let scale = n_cells as f64 / box_size;
    let ix = ((point[0] * scale).floor() as i32).rem_euclid(n_cells);
    let iy = ((point[1] * scale).floor() as i32).rem_euclid(n_cells);
    let iz = ((point[2] * scale).floor() as i32).rem_euclid(n_cells);
    (ix, iy, iz)
}

fn neighbor_cells(cell: (i32, i32, i32), n_cells: i32) -> Vec<(i32, i32, i32)> {
    let mut result = Vec::with_capacity(27);
    for dx in -1..=1 {
        for dy in -1..=1 {
            for dz in -1..=1 {
                result.push((
                    (cell.0 + dx).rem_euclid(n_cells),
                    (cell.1 + dy).rem_euclid(n_cells),
                    (cell.2 + dz).rem_euclid(n_cells),
                ));
            }
        }
    }
    result
}

/// cutoff / cell list の正規化
pub fn normalize_neighbor_options(
    box_size: f64,
    use_cutoff: bool,
    cutoff_radius: Option<f64>,
    use_cell_list: bool,
    cell_size: Option<f64>,
) -> Result<(bool, Option<f64>, bool, Option<f64>), String> {
    let effective_use_cutoff = use_cutoff;
    let mut effective_cutoff = cutoff_radius;
    let effective_use_cell_list = use_cutoff && use_cell_list;
    let mut effective_cell_size = cell_size;

    if effective_use_cutoff {
        let rc = effective_cutoff.ok_or("cutoff_radius must be provided when use_cutoff=true")?;
        if rc <= 0.0 {
            return Err(format!("cutoff_radius must be positive, got {}", rc));
        }
        if rc >= 0.5 * box_size {
            return Err(format!(
                "cutoff_radius must be smaller than half the box size. Got {}",
                rc
            ));
        }
    } else {
        effective_cutoff = None;
    }

    if effective_use_cell_list {
        if effective_cell_size.is_none() {
            effective_cell_size = effective_cutoff;
        }
        if effective_cell_size.map_or(true, |s| s <= 0.0) {
            return Err("cell_size must be positive".to_string());
        }
    } else {
        effective_cell_size = None;
    }

    Ok((effective_use_cutoff, effective_cutoff, effective_use_cell_list, effective_cell_size))
}

/// 近傍探索状態を構築
pub fn build_neighbor_state(
    pos: &[Vec3],
    box_size: f64,
    use_cutoff: bool,
    cutoff_radius: Option<f64>,
    use_cell_list: bool,
    cell_size: Option<f64>,
) -> Result<NeighborState, String> {
    let (_, eff_cutoff, eff_use_cl, eff_cell_size) =
        normalize_neighbor_options(box_size, use_cutoff, cutoff_radius, use_cell_list, cell_size)?;

    let mut state = NeighborState::new(use_cutoff, eff_cutoff, eff_use_cl, eff_cell_size);

    if !eff_use_cl {
        return Ok(state);
    }

    let cell_size_val = eff_cell_size.unwrap();
    let n_cells = (box_size / cell_size_val).floor().max(1.0) as i32;
    state.n_cells = Some(n_cells);
    state.cells = Some(HashMap::new());
    state.particle_cells = Some(Vec::with_capacity(pos.len()));

    for (idx, &point) in pos.iter().enumerate() {
        let cell = cell_index(point, box_size, n_cells);
        state.particle_cells.as_mut().unwrap().push(cell);
        state
            .cells
            .as_mut()
            .unwrap()
            .entry(cell)
            .or_default()
            .push(idx);
    }

    Ok(state)
}

/// 1粒子変位後の cell list 差分更新
pub fn update_neighbor_state_after_move(
    state: &mut NeighborState,
    idx: usize,
    old_pos: Vec3,
    new_pos: Vec3,
    box_size: f64,
) {
    if state.cells.is_none()
        || state.particle_cells.is_none()
        || state.n_cells.is_none()
    {
        return;
    }

    let n_cells = state.n_cells.unwrap();
    let old_cell = cell_index(old_pos, box_size, n_cells);
    let new_cell = cell_index(new_pos, box_size, n_cells);

    if old_cell == new_cell {
        return;
    }

    let cells = state.cells.as_mut().unwrap();
    if let Some(list) = cells.get_mut(&old_cell) {
        list.retain(|&i| i != idx);
        if list.is_empty() {
            cells.remove(&old_cell);
        }
    }
    cells.entry(new_cell).or_default().push(idx);
    state.particle_cells.as_mut().unwrap()[idx] = new_cell;
}

/// 同じ設定で近傍探索状態を再構築（粒子追加・削除後用）
pub fn rebuild_neighbor_state(
    state: Option<&NeighborState>,
    pos: &[Vec3],
    box_size: f64,
) -> Option<NeighborState> {
    let s = state?;
    build_neighbor_state(
        pos,
        box_size,
        s.use_cutoff,
        s.cutoff_radius,
        s.use_cell_list,
        s.cell_size,
    )
    .ok()
}

fn candidate_indices_for_particle(
    pos: &[Vec3],
    i: usize,
    _box_size: f64,
    state: &NeighborState,
) -> Vec<usize> {
    if state.cells.is_none()
        || state.particle_cells.is_none()
        || state.n_cells.is_none()
    {
        return (0..pos.len()).collect();
    }

    let cell = state.particle_cells.as_ref().unwrap()[i];
    let n_cells = state.n_cells.unwrap();
    let cells = state.cells.as_ref().unwrap();
    let mut seen = HashSet::new();
    for neighbor in neighbor_cells(cell, n_cells) {
        if let Some(list) = cells.get(&neighbor) {
            for &idx in list {
                seen.insert(idx);
            }
        }
    }
    seen.into_iter().collect()
}

fn candidate_indices_for_point(
    candidate: Vec3,
    pos: &[Vec3],
    box_size: f64,
    state: &NeighborState,
) -> Vec<usize> {
    if state.cells.is_none() || state.n_cells.is_none() {
        return (0..pos.len()).collect();
    }

    let n_cells = state.n_cells.unwrap();
    let cell = cell_index(candidate, box_size, n_cells);
    let cells = state.cells.as_ref().unwrap();
    let mut seen = HashSet::new();
    for neighbor in neighbor_cells(cell, n_cells) {
        if let Some(list) = cells.get(&neighbor) {
            for &idx in list {
                seen.insert(idx);
            }
        }
    }
    seen.into_iter().collect()
}

/// 粒子 i と他粒子の LJ 相互作用エネルギー
pub fn calc_energy(
    pos: &[Vec3],
    i: usize,
    box_size: f64,
    epsilon: f64,
    sigma: f64,
    state: &NeighborState,
) -> f64 {
    let use_cutoff = state.use_cutoff;
    let cutoff_radius = state.cutoff_radius;

    let mut energy = 0.0;
    for &j in &candidate_indices_for_particle(pos, i, box_size, state) {
        if i == j {
            continue;
        }
        let delta = delta_minimum_image(pos[i], pos[j], box_size);
        let r = norm(delta);
        energy += lj_pair_energy(r, epsilon, sigma, use_cutoff, cutoff_radius);
    }
    energy
}

/// 全粒子系の全エネルギー
pub fn total_energy(
    pos: &[Vec3],
    box_size: f64,
    epsilon: f64,
    sigma: f64,
    state: &NeighborState,
) -> f64 {
    let mut sum = 0.0;
    for i in 0..pos.len() {
        sum += calc_energy(pos, i, box_size, epsilon, sigma, state);
    }
    sum / 2.0
}

/// ランダム初期配置（min_distance 以上離して配置）
pub fn init_positions(
    box_size: f64,
    n_particles: usize,
    min_distance: f64,
    rng: &mut impl Rng,
    max_trials: usize,
    max_global_retries: usize,
) -> Result<Vec<Vec3>, String> {
    let min_distance_sq = min_distance * min_distance;

    for _ in 0..max_global_retries {
        let mut pos = Vec::with_capacity(n_particles);

        for idx in 0..n_particles {
            let mut placed = false;
            for _ in 0..max_trials {
                let candidate: Vec3 = [
                    rng.gen_range(0.0..box_size),
                    rng.gen_range(0.0..box_size),
                    rng.gen_range(0.0..box_size),
                ];

                if idx == 0 {
                    pos.push(candidate);
                    placed = true;
                    break;
                }

                let mut ok = true;
                for &p in &pos {
                    let delta = delta_minimum_image(candidate, p, box_size);
                    let r_sq = delta[0] * delta[0] + delta[1] * delta[1] + delta[2] * delta[2];
                    if r_sq < min_distance_sq {
                        ok = false;
                        break;
                    }
                }
                if ok {
                    pos.push(candidate);
                    placed = true;
                    break;
                }
            }
            if !placed {
                break;
            }
        }
        if pos.len() == n_particles {
            return Ok(pos);
        }
    }

    Err(format!(
        "Failed to place {} particles with min_distance={} in box_size={}",
        n_particles, min_distance, box_size
    ))
}

/// 候補位置に1粒子を挿入したときの相互作用エネルギー
pub fn insertion_energy(
    pos: &[Vec3],
    candidate: Vec3,
    box_size: f64,
    epsilon: f64,
    sigma: f64,
    state: &NeighborState,
) -> f64 {
    if pos.is_empty() {
        return 0.0;
    }

    let use_cutoff = state.use_cutoff;
    let cutoff_radius = state.cutoff_radius;

    let mut energy = 0.0;
    for &idx in &candidate_indices_for_point(candidate, pos, box_size, state) {
        let delta = delta_minimum_image(candidate, pos[idx], box_size);
        let r = norm(delta);
        energy += lj_pair_energy(r, epsilon, sigma, use_cutoff, cutoff_radius);
    }
    energy
}
